# 回帰

### データの読み込み

In [2]:
"""読み込み
"""
from sklearn.datasets import load_boston
boston = load_boston()

In [3]:
"""データの内訳
"""
print(boston)

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]]), 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 1

### データ分割
訓練データ：モデルを作成するためのデータ

テストデータ：作成したモデルの性能を検証するデータ

過学習：訓練データにのみ当てはまりの良いモデルを作成してしまうこと。訓練データ・テストデータ双方に当てはまりがよくないといいモデルとは言えない

In [90]:
"""データ分割
・説明変数と目的変数を与える
・テストデータを全体の何％にするか決める
・固定にするか否か
"""
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(boston['data'], boston['target'], test_size=0.3,  random_state=0)

In [91]:
"""基本モデルの作成
"""
from sklearn.neural_network import MLPRegressor
clf = MLPRegressor()

In [61]:
"""標準化
"""
import scipy.stats
X_train=scipy.stats.zscore(X_train)
X_test=scipy.stats.zscore(X_test)
#y_train=scipy.stats.zscore(y_train)
#y_test=scipy.stats.zscore(y_test)

In [92]:
"""モデルの最適化
"""
clf.fit(X_train, y_train)

c:\users\mio\appdata\local\programs\python\python36-32\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [93]:
"""テストデータの予測
"""
predicted = clf.predict(X_test)

In [94]:
"""予測値
"""
print(predicted)

[25.63596252 24.1591701  28.88816885 15.83107205 18.3299765  22.96346937
 22.52452177 26.00629923 17.49987321  4.81086808  8.750776   12.68292742
 14.48325325 10.25317173 31.73952025 29.40801372 22.89600272 32.34543531
 25.46970732 23.41504343 24.40856269 21.58486732 20.95722693 28.86759611
 23.91855215 13.2041683  22.79483738 15.4323776  30.60598126 20.58078999
 17.75563422 19.91272171 22.4972579  24.31104861 27.32468768 19.14175978
 11.35922247 22.9734945  12.13588686  9.23095269 25.28947293 21.7993401
 22.79686049 15.89587876 20.68298237 28.10627785 18.28808618 19.57478495
 12.91757002 23.49279192  5.74005378 19.46659416 20.78480502 27.94529273
 13.81189672 18.84433326 22.95822279 15.54925183 16.62136005 23.41156898
 27.39741368 20.99211852 27.61032046 26.16016744 20.39446459 28.6350141
 17.23566292 19.46106945 11.42289712 24.15669592 19.34659949 20.94355212
 30.73193286 28.99407353 23.93947562  7.89542536 26.99416327 22.31327892
 25.91693609 19.09481178 24.38747363 22.9439128  19.7

In [95]:
"""正解率の算出
乖離度　=　Σ(|予測値 - 実測値|/実測値)/N
N：データ数
"""
print(
    sum(abs(predicted-y_test)/y_test)/len(y_test)
)

0.21771199166984562


# データの読み込み
irisデータセットという「花（アヤメ）」に関する有名なデータがあります

data：ガクの長さ・ガクの幅・花弁の長さ・花弁の幅

target：花（アヤメ）の種類

In [15]:
"""読み込み
"""
from sklearn.datasets import load_iris
iris = load_iris()

In [16]:
"""データの内訳
"""
print(iris)

{'data': array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
     

# データ分割
訓練データ：モデルを作成するためのデータ

テストデータ：作成したモデルの性能を検証するデータ

過学習：訓練データにのみ当てはまりの良いモデルを作成してしまうこと。訓練データ・テストデータ双方に当てはまりがよくないといいモデルとは言えない

In [84]:
"""データ分割
・説明変数と目的変数を与える
・テストデータを全体の何％にするか決める
・固定にするか否か
"""
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(iris['data'], iris['target'], test_size=0.3,  random_state=0)

In [85]:
"""基本モデルの作成
"""
from sklearn.neural_network import MLPClassifier
clf = MLPClassifier()

In [86]:
"""モデルの最適化
"""
clf.fit(X_train, y_train)

c:\users\mio\appdata\local\programs\python\python36-32\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [87]:
"""テストデータの予測
"""
predicted = clf.predict(X_test)

In [88]:
"""正解率の算出
"""
print(sum(predicted==y_test)/len(y_test))

0.9555555555555556


# 各種パラメータ

|パラメータ名|デフォルト|意味・使い方|取りうる値|
|:---|:---|:---|:---|
|hidden_layer_sizes|(100,)|隠れ層とノード数||
|activation|"relu"|活性化関数|"identity"・"logistic"・"tanh"・"relu"|
|solver|"adam"|最適化手法|"lbfgs"・"sgd"・"adam"|
|alpha|0.0001|L2正則化|小さくしすぎると過学習になる|
|batch_size|min(200, n_samples)|ミニバッチのサイズ||
|learning_rate_init|?|学習率の初期値||
|learning_rate|"constant"|学習率の更新の仕方|solver="sgd"の時のみ有効|
|power_t|0.5|学習率の減少速度を指定|solver="sgd" and learning_rate="invscaling"の時のみ有効|
|max_iter|200|学習回数|intで指定。途中で学習が収束したと判断した場合は自動的に終了|
|shuffle|True|学習ごとにデータをシャッフルするか|solver="sgd" or solver="adam"の時のみ有効|
|random_state|None|実行ごとに結果を変化させないようにする|int|
|tol|1e-4|計算停止の閾値|not solver="adaptive"の時のみ有効|
|verbose|False|経過を表示するか||
|warm_start|False|学習済みモデルの再学習をするか否か||
|momentum|0.9|学習係数|solver="sgd"の時のみ有効|
|nesterovs_momentum|True|勾配計算時にmomentum考慮するか否か|solver="sgd" and momentum>0の時のみ有効|
|early_stopping|False|早期停止するか否か|基準は2回連続でvalidation_fractionでの計算結果がtolを下回る|
|validation_fraction|0.1|early_stoppingで使用するデータ数|floatで指定|
|beta_1|0.9|係数|solver="adam"の時のみ有効|
|beta_2|0.999|係数|solver="adam"の時のみ有効|
|epsilon|1e-8|係数|solver="adam"の時のみ有効|

In [66]:
"""基本モデルの作成
"""
X_train, X_test, y_train, y_test = train_test_split(boston['data'], boston['target'], test_size=0.3,  random_state=0)
clf=MLPRegressor(
    hidden_layer_sizes=(100,200,200,50,),
    solver="sgd",
    max_iter=300,
    momentum=0.85
)

In [67]:
"""標準化
"""
import scipy.stats
X_train=scipy.stats.zscore(X_train)
X_test=scipy.stats.zscore(X_test)
#y_train=scipy.stats.zscore(y_train)
#y_test=scipy.stats.zscore(y_test)

In [68]:
"""モデルの最適化
"""
clf.fit(X_train, y_train)

c:\users\mio\appdata\local\programs\python\python36-32\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100, 200, 200, 50), learning_rate='constant',
       learning_rate_init=0.001, max_iter=300, momentum=0.85,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='sgd', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [69]:
"""テストデータの予測
"""
predicted = clf.predict(X_test)

In [70]:
"""予測値
"""
print(predicted)

[22.27592036 29.23833312 25.68734163 11.23324153 20.56994372 20.47264713
 24.55712675 21.32691515 18.9952227  17.84591488  9.49342607 12.91661729
 16.44999044 10.09077743 48.44464469 37.0812037  24.79002289 40.3258392
 31.77996192 22.64881087 24.39107557 22.56638514 21.34311908 28.1193196
 22.15762195 24.70065553 17.33748681 19.270581   41.05593288 16.98396772
 15.25183041 18.64898317 19.50509003 20.51635558 28.10571455 22.05257429
  7.87301051 40.26418661 15.58128428 14.64457343 25.37350998 24.17995178
 19.15304857 19.03002351 18.55823721 25.6211893  21.64658523 15.68752589
 16.19703339 23.31575105 15.25980374 23.80496682 24.27441464 47.59797285
 14.51300727 21.85829954 17.71365709 18.44705238 20.35519936 20.61527145
 21.05361264 22.36959384 35.82234636 33.15367447 19.86962885 31.16026942
 17.35054378 21.08288862 14.29277006 24.63317631 21.29119394 22.26975725
 27.5194465  31.04155909 26.54604241  9.31502823 43.96277466 24.26501017
 27.44455746 18.82242395 27.16859077 20.1639179  24.9

In [71]:
"""正解率の算出
乖離度　=　Σ(|予測値 - 実測値|/実測値)/N
N：データ数
"""
print(
    sum(abs(predicted-y_test)/y_test)/len(y_test)
)

0.12922271564432128
